internal/logger/io_sink.go

```diff
@@ -7,91 +7,43 @@
-	"log"
+	"time"
 )
 
-// IOSink writes to an io.Writer using the standard library logging solution and
-// is the default sink for the logger, with the default IO being os.Stderr.
+// IOSink writes a JSON-encoded message to the io.Writer.
 type IOSink struct {
-	log *log.Logger
+	enc *json.Encoder
 }
 
-// Compile-time check to ensure osSink implements the LogSink interface.
+// Compile-time check to ensure IOSink implements the LogSink interface.
 var _ LogSink = &IOSink{}
 
-// NewIOSink will create a new IOSink that writes to the provided io.Writer.
+// NewIOSink will create an IOSink object that writes JSON messages to the
+// provided io.Writer.
 func NewIOSink(out io.Writer) *IOSink {
 	return &IOSink{
-		log: log.New(out, "", log.LstdFlags),
+		enc: json.NewEncoder(out),
 	}
 }
 
-func logCommandMessageStarted(log *log.Logger, kvMap map[string]interface{}) {
-	format := "Command %q started on database %q using a connection with " +
-		"server-generated ID %d to %s:%d. The requestID is %d and " +
-		"the operation ID is %d. Command: %s"
+// Info will write a JSON-encoded message to the io.Writer.
+func (sink *IOSink) Info(_ int, msg string, keysAndValues ...interface{}) {
+	kvMap := make(map[string]interface{}, len(keysAndValues)/2+2)
 
-	log.Printf(format,
-		kvMap["commandName"],
-		kvMap["databaseName"],
-		kvMap["serverConnectionId"],
-		kvMap["serverHost"],
-		kvMap["serverPort"],
-		kvMap["requestId"],
-		kvMap["operationId"],
-		kvMap["command"])
+	kvMap[KeyTimestamp] = time.Now().UnixNano()
+	kvMap[KeyMessage] = msg
 
-}
-
-func logCommandMessageSucceeded(log *log.Logger, kvMap map[string]interface{}) {
-	format := "Command %q succeeded in %d ms using server-generated ID " +
-		"%d to %s:%d. The requestID is %d and the operation ID is " +
-		"%d. Command reply: %s"
-
-	log.Printf(format,
-		kvMap["commandName"],
-		kvMap["duration"],
-		kvMap["serverConnectionId"],
-		kvMap["serverHost"],
-		kvMap["serverPort"],
-		kvMap["requestId"],
-		kvMap["operationId"],
-		kvMap["reply"])
-}
-
-func logCommandMessageFailed(log *log.Logger, kvMap map[string]interface{}) {
-	format := "Command %q failed in %d ms using a connection with " +
-		"server-generated ID %d to %s:%d. The requestID is %d and " +
-		"the operation ID is %d. Error: %s"
-
-	log.Printf(format,
-		kvMap["commandName"],
-		kvMap["duration"],
-		kvMap["serverConnectionID"],
-		kvMap["serverHost"],
-		kvMap["serverPort"],
-		kvMap["requestId"],
-		kvMap["operationId"],
-		kvMap["failure"])
-}
-
-func (osSink *IOSink) Info(_ int, msg string, keysAndValues ...interface{}) {
-	kvMap := make(map[string]interface{})
 	for i := 0; i < len(keysAndValues); i += 2 {
 		kvMap[keysAndValues[i].(string)] = keysAndValues[i+1]
 	}
 
-	switch msg {
-	case CommandStarted:
-		logCommandMessageStarted(osSink.log, kvMap)
-	case CommandSucceeded:
-		logCommandMessageSucceeded(osSink.log, kvMap)
-	case CommandFailed:
-		logCommandMessageFailed(osSink.log, kvMap)
-	}
+	_ = sink.enc.Encode(kvMap)
```

Comment:
I just realized that a `json.Encoder` isn't documented as safe for concurrent use, so we need to synchronize this call (i.e. add a mutex or some other lock).

Reply:
It might be better to add a mutex to the logger.Print method. The only reason it makes sense to add a mutex at all is that the Driver might make asynchronous calls to logger.Print. If that's the concern, we should guard there to protect user-all usages of a Sink and not just our IOSink.